pkgs/fc/ceph/src/fc/ceph/osd/nautilus.py

```diff
@@ -469,26 +487,61 @@ def _create_crush_and_auth(self, volume, crush_location):
 
         run.ceph("osd", "crush", "add", self.name, str(weight), crush_location)
 
-    def purge(self, unsafe_destroy):
-        """Deletes an osd, including removal of auth keys and crush map entry"""
+    @staticmethod
+    def run_safety_check(ids: List[int], strict_safety_check: bool):
```

Comment:
We could turn this into a decorator to simplify the individual use cases so that this code could be a wrapper around the protected actions

```
@safety_check
def destroy(...)
   ...

def safety_check(f):
    def wrapped(self, *args, **, no_safety_check, strict_safety_check, **kw):
        # ... do the safety check here
        f(self, *args, **kw)
    return wrapped
```

There are specific helpers in the stdlib to assist in the wrapping (I think there's a decorator decorator) so that the wrapped methods get proper naming, etc.

Reply:
I'll have a look at them. One reason to keep the check function rather short and focused was that I wanted to adjust the particular warning texts anyways. I could probably give them as a paramteter for the decorater, which would turn its code footprint rather huge though.

Reply:
The issue here is that both the safety check as well as the destroy/deactivate functions require the osd ID or list of osd IDs as an argument.
I do not immediately see a simple way to refactor this in a non-hacky way (e.g. accessing the kwargs of the wrapped function) and thus do not see the effort being worth it. Especially as the usage semantics between destroy and deactivate are slightly different as well.